edgedb-protocol/src/descriptors.rs

```diff
@@ -490,29 +546,96 @@ impl Descriptor {
+impl<T: Decode> Decode for Vec<T> {
+    fn decode(buf: &mut Input) -> Result<Self, DecodeError> {
+        ensure!(buf.remaining() >= 2, errors::Underflow);
+        let element_count = buf.get_u16();
+        let mut elements = Vec::with_capacity(element_count as usize);
+        for _ in 0..element_count {
+            elements.push(T::decode(buf)?);
+        }
+        Ok(elements)
+    }
+}
+
+impl Decode for Option<u32> {
+    fn decode(buf: &mut Input) -> Result<Self, DecodeError> {
+        ensure!(buf.remaining() >= 4, errors::Underflow);
+
+        let val = match buf.get_i32() {
+            -1 => None,
+            n if n > 0 => Some(n as u32),
+            _ => errors::InvalidOptionU32.fail()?,
+        };
+
+        Ok(val)
+    }
+}
+
+impl Decode for TypePos {
+    fn decode(buf: &mut Input) -> Result<Self, DecodeError> {
+        ensure!(buf.remaining() >= 2, errors::Underflow);
+        Ok(Self(buf.get_u16()))
+    }
 }
 
 impl Decode for Descriptor {
     fn decode(buf: &mut Input) -> Result<Self, DecodeError> {
         use Descriptor as D;
+        if buf.proto().is_at_least(2, 0) {
+            ensure!(buf.remaining() >= 4, errors::Underflow);
+            let desc_len = buf.get_u32() as u64;
+            ensure!((buf.remaining() as u64) >= desc_len, errors::Underflow);
+        }
         ensure!(buf.remaining() >= 1, errors::Underflow);
         match buf.chunk()[0] {
-            0 => SetDescriptor::decode(buf).map(D::Set),
-            1 => ObjectShapeDescriptor::decode(buf).map(D::ObjectShape),
-            2 => BaseScalarTypeDescriptor::decode(buf).map(D::BaseScalar),
-            3 => ScalarTypeDescriptor::decode(buf).map(D::Scalar),
-            4 => TupleTypeDescriptor::decode(buf).map(D::Tuple),
-            5 => NamedTupleTypeDescriptor::decode(buf).map(D::NamedTuple),
-            6 => ArrayTypeDescriptor::decode(buf).map(D::Array),
-            7 => EnumerationTypeDescriptor::decode(buf).map(D::Enumeration),
-            8 => InputShapeTypeDescriptor::decode(buf).map(D::InputShape),
-            9 => RangeTypeDescriptor::decode(buf).map(D::Range),
+            0x00 => SetDescriptor::decode(buf).map(D::Set),
+            0x01 => ObjectShapeDescriptor::decode(buf).map(D::ObjectShape),
+            0x02 => BaseScalarTypeDescriptor::decode(buf).map(D::BaseScalar),
+            0x03 => ScalarTypeDescriptor::decode(buf).map(D::Scalar),
+            0x04 => TupleTypeDescriptor::decode(buf).map(D::Tuple),
+            0x05 => NamedTupleTypeDescriptor::decode(buf).map(D::NamedTuple),
+            0x06 => ArrayTypeDescriptor::decode(buf).map(D::Array),
+            0x07 => EnumerationTypeDescriptor::decode(buf).map(D::Enumeration),
+            0x08 => InputShapeTypeDescriptor::decode(buf).map(D::InputShape),
+            0x09 => RangeTypeDescriptor::decode(buf).map(D::Range),
+            0x0A => ObjectTypeDescriptor::decode(buf).map(D::Object),
+            0x0B => CompoundTypeDescriptor::decode(buf).map(D::Compound),
```

Comment:
Ideally these should be constants, perhaps attached to the `XXXTypeDescriptor` as a trait.

Reply:
Yeah, can be done in a followup.